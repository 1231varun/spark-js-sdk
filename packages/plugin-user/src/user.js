/**!
 *
 * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
 * @private
 */

import {deprecated} from 'core-decorators';
import {isArray} from 'lodash';
import {oneFlight, patterns, tap} from '@ciscospark/common';
import {SparkPlugin} from '@ciscospark/spark-core';
import UserUUIDBatcher from './user-uuid-batcher';
import UserUUIDStore from './user-uuid-store';

const User = SparkPlugin.extend({
  namespace: `User`,

  children: {
    batcher: UserUUIDBatcher
  },

  session: {
    store: {
      default() {
        return new UserUUIDStore();
      },
      type: `any`
    }
  },

  /**
   * Activates a Spark user account and exchanges for user token.
   * @param {Object} options
   * @param {Object} options.verificationToken (required)
   * @returns {Promise} Resolves with a userSession
   */
  activate(options) {
    options = options || {};

    if (!options.verificationToken) {
      throw new Error(`\`options.verificationToken\` is required`);
    }
    let response;

    return this.request({
      uri: this.config.activationUrl,
      method: `POST`,
      body: options,
      auth: {
        user: this.spark.config.credentials.oauth.client_id,
        pass: this.spark.config.credentials.oauth.client_secret,
        sendImmediately: true
      },
      withCredentials: true
    })
      .then((res) => {
        response = res;
        return this.getOauthCode();
      })
      .then((res) => {
        const code = res.match(`<title>(.*)</title>`)[1];
        return this.spark.credentials.requestAuthorizationCodeGrant({code});
      })
      .then(() => {
        this.setPasswordStatus(false);
        return response.body;
      });
  },

  /**
   * Converts a user-identifying object to a uuid, perhaps by doing a network
   * lookup
   * @param {string|Object} user
   * @param {Object} options
   * @param {boolean} options.create if true, ensures the return UUID refers to
   * an existing user (rather than creating one deterministically based on email
   * address), even if that user must be created
   * @returns {Promise<string>}
   */
  asUUID(user, options) {
    if (!user) {
      return Promise.reject(new Error(`\`user\` is required`));
    }

    if (isArray(user)) {
      return Promise.all(user.map((u) => this.asUUID(u, options)));
    }

    const id = this._extractUUID(user);
    if (!(options && options.force) && patterns.uuid.test(id)) {
      return Promise.resolve(id);
    }

    const email = this._extractEmailAddress(user);

    if (!patterns.email.test(email)) {
      return Promise.reject(new Error(`Provided user object does not appear to identify a user`));
    }

    return this.getUUID(email, options);
  },

  /**
   * Requests a uuid from the api
   * @param {string} email
   * @param {Object} options
   * @param {boolean} options.create
   * @returns {Promise<string>}
   */
  fetchUUID(email, options) {
    return this.batcher.request({
      email,
      create: options && options.create
    })
      .then((user) => this.recordUUID(Object.assign({emailAddress: email}, user))
        .then(() => user.id));
  },

  /**
   * Fetches details about the current user
   * @returns {Promise<Object>}
   */
  get() {
    return this.request({
      service: `conversation`,
      resource: `users`
    })
      .then((res) => res.body)
      .then(tap((user) => this.recordUUID(user)));
  },

  /**
   * Converts an email address to a uuid, perhaps by doing a network lookup
   * @param {string} email
   * @param {Object} options
   * @param {boolean} options.create
   * @returns {Promise<string>}
   */
  @oneFlight({keyFactory: (email, options) => email + String(options && options.create)})
  getUUID(email, options) {
    return this.store.getByEmail(email)
      .then((user) => {
        if (options && options.create && !user.userExists) {
          return Promise.reject(new Error(`User for specified email cannot be confirmed to exist`));
        }

        if (!user.id) {
          return Promise.reject(new Error(`No id recorded for specified user`));
        }

        return user.id;
      })
      .catch(() => this.fetchUUID(email, options));
  },

  /**
   * Caches the uuid for the specified email address
   * @param {Object} user
   * @param {string} user.id
   * @param {string} user.emailAddress
   * @returns {Promise}
   */
  recordUUID(user) {
    if (!user) {
      return Promise.reject(new Error(`\`user\` is required`));
    }

    if (!user.id) {
      return Promise.reject(new Error(`\`user.id\` is required`));
    }

    if (!patterns.uuid.test(user.id)) {
      return Promise.reject(new Error(`\`user.id\` must be a uuid`));
    }

    if (!user.emailAddress) {
      return Promise.reject(new Error(`\`user.emailAddress\` is required`));
    }

    if (!patterns.email.test(user.emailAddress)) {
      return Promise.reject(new Error(`\`user.emailAddress\` must be an email address`));
    }

    return this.store.add(user);
  },

  @deprecated(`Use User#verify()`)
  register(...args) {
    return this.verify(...args);
  },

  /**
   * Sets password status on token
   * @param {Boolean} value
   * @returns {undefined}
   */
  setPasswordStatus(value) {
    if (this.spark.credentials && this.spark.credentials.supertoken) {
      this.spark.credentials.supertoken.passwordSet = value;
    }
  },

  /**
   * Updates a user's password with spark.
   * @param {Object} options
   * @param {string} options.password (required)
   * @param {string} options.userId (required)
   * @returns {Promise} Resolves with complete user object containing new password
   */
  setPassword(options) {
    options = options || {};
    if (!options.password) {
      return Promise.reject(new Error(`\`options.password\` is required`));
    }
    if (!options.userId) {
      return Promise.reject(new Error(`\`options.userId\` is required`));
    }

    let headers;

    const promise = this.spark.credentials.getAuthorization()
      .then((authorization) => {
        headers = {authorization};
      });

    return promise
      .then(() => this.request({
        uri: `${this.config.setPasswordUrl}/${options.userId}`,
        method: `PATCH`,
        headers,
        body: {
          schemas: [`urn:scim:schemas:core:1.0`, `urn:scim:schemas:extension:cisco:commonidentity:1.0`],
          password: options.password
        }
      }))
      .then(() => {
        this.setPasswordStatus(true);
      });
  },

  /**
   * Updates the current user's display name
   * @param {Object} options
   * @param {string} options.displayName
   * @returns {Promise<Object>}
   */
  update(options) {
    if (!options.displayName) {
      return Promise.reject(new Error(`\`options.displayName\` is required`));
    }

    return this.request({
      method: `PATCH`,
      service: `conversation`,
      resource: `users/user`,
      body: options
    })
      .then((res) => res.body);
  },

  /**
   * Determines if the speicfied user needs to signup or can signin.
   * Triggers activation email if client credentials are used
   * @param {Object} options
   * @param {string} options.email (required)
   * @param {string} options.reqId required for the endpoint
   * @returns {Promise<Object>}
   */
  verify(options) {
    options = Object.assign({}, this.config.verifyDefaults, options);

    if (!options.email) {
      return Promise.reject(new Error(`\`options.email\` is required`));
    }

    if (!options.reqId) {
      return Promise.reject(new Error(`\`options.reqId\` is required`));
    }

    let shouldRefreshAccessToken = true;
    let requiresClientCredentials = false;
    let headers;
    const promise = this.spark.credentials.getAuthorization()
      .then((authorization) => {
        headers = {authorization};
      })
      .catch(() => this.spark.credentials.getClientCredentialsAuthorization()
        .then((authorization) => {
          headers = {authorization};
          shouldRefreshAccessToken = false;
          requiresClientCredentials = true;
        })
        .catch((err) => Promise.reject(new Error(`failed to set authorization`, err)))
      );

    return promise
      .then(() => this.request({
        service: `atlas`,
        resource: `users/activations`,
        method: `POST`,
        headers,
        body: options,
        shouldRefreshAccessToken,
        requiresClientCredentials
      }))
      .then((res) => res.body);
  },

  /**
   * Extracts the uuid from a user identifying object
   * @param {string|Object} user
   * @private
   * @returns {string}
   */
  _extractUUID(user) {
    return user.entryUUID || user.id || user;
  },

  /**
   * Extracts the email address from a user identifying object
   * @param {string|Object} user
   * @private
   * @returns {string}
   */
  _extractEmailAddress(user) {
    return user.email || user.emailAddress || user.entryEmail || user;
  },

  /**
   * Uses cookie header to request quth code
   * @private
   * @returns {string} html body with auth code
   */
  _getOauthCode() {
    return this.request({
      api: `oauth`,
      withCredentials: true,
      method: `POST`,
      resource: `authorize`,
      form: {
        /* eslint camelcase: [0] */
        response_type: `code`,
        redirect_uri: `urn:ietf:wg:oauth:2.0:oob`,
        client_id: this.spark.config.credentials.oauth.client_id,
        scope: this.spark.config.credentials.oauth.scope,
        service: `spark`
      }
    })
      .then((res) => res.body);
  }
});

export default User;
