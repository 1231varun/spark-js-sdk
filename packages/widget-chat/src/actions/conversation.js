import {updateHasNewMessage} from './widget.js';
import {setTyping} from './indicators';

export const ADD_ACTIVITIES_TO_CONVERSATION = `ADD_ACTIVITIES_TO_CONVERSATION`;
export function addActivitiesToConversation(activities) {
  return {
    type: ADD_ACTIVITIES_TO_CONVERSATION,
    payload: {
      activities
    }
  };
}

export const CREATE_CONVERSATION_BEGIN = `CREATE_CONVERSATION_BEGIN`;
export function createConversationBegin(userId) {
  return {
    type: CREATE_CONVERSATION_BEGIN,
    payload: {
      userId
    }
  };
}

export const CREATE_CONVERSATION = `CREATE_CONVERSATION`;
export function createConversation(conversation) {
  return {
    type: CREATE_CONVERSATION,
    payload: {
      conversation
    }
  };
}

export const DELETE_ACTIVITY_FROM_CONVERSATION = `DELETE_ACTIVITY_FROM_CONVERSATION`;
export function deleteActivityFromConversation(conversation, activity) {
  return {
    type: DELETE_ACTIVITY_FROM_CONVERSATION,
    payload: {
      conversation,
      activity
    }
  };
}

export const RECEIVE_MERCURY_COMMENT = `RECEIVE_MERCURY_COMMENT`;
export function receiveMercuryComment(activity) {
  return {
    type: RECEIVE_MERCURY_COMMENT,
    payload: {
      activity
    }
  };
}

export const RECEIVE_MERCURY_ACTIVITY = `RECEIVE_MERCURY_ACTIVITY`;
export function receiveMercuryActivity(activity) {
  return {
    type: RECEIVE_MERCURY_ACTIVITY,
    payload: {
      activity
    }
  };
}

export const UPDATE_MERCURY_STATE = `UPDATE_MERCURY_STATE`;
export function updateMercuryState(mercuryState) {
  return {
    type: UPDATE_MERCURY_STATE,
    payload: {
      mercuryState
    }
  };
}

export const UPDATE_CONVERSATION_STATE = `UPDATE_CONVERSATION_STATE`;
export function updateConversationState(conversationState) {
  return {
    type: UPDATE_CONVERSATION_STATE,
    payload: {
      conversationState
    }
  };
}


/**
 * Creates/Opens a conversation with a user
 *
 * @param {String} userId Either UUID or email of user
 * @param {object} spark
 * @returns {function}
 */
export function createConversationWithUser(userId, spark) {
  return (dispatch) => {
    dispatch(createConversationBegin(userId));

    spark.conversation.create({
      participants: [userId]
    }, {
      latestActivity: true,
      activitiesLimit: 30
    })
      .then((conversation) => dispatch(createConversation(conversation)));
  };
}

export function deleteActivity(conversation, activity, spark) {
  return (dispatch) =>
    spark.conversation.delete(conversation, activity)
      .then(() => {
        dispatch(deleteActivityFromConversation(conversation, activity));
      });
}

export function listenToMercuryActivity(conversationId, spark) {
  return (dispatch) => {
    dispatch(updateMercuryState({isListening: true}));
    spark.mercury.on(`event:status.start_typing`, (event) => {
      if (event.data.conversationId === conversationId) {
        dispatch(setTyping(event.data.actor.id, true));
      }
    });

    spark.mercury.on(`event:status.stop_typing`, (event) => {
      if (event.data.conversationId === conversationId) {
        dispatch(setTyping(event.data.actor.id, false));
      }
    });

    spark.mercury.on(`event:conversation.activity`, (event) => {
      const activity = event.data.activity;
      const isChatMessage = activity.verb === `post` && activity.object.objectType === `comment` && activity.target.objectType === `conversation`;
      // Ignore activity from other conversations
      if (activity.target.id === conversationId) {
        if (isChatMessage) {
          dispatch(updateHasNewMessage(true));
          dispatch(receiveMercuryComment(activity));
        }
        else if (activity.object.objectType === `activity`) {
          dispatch(receiveMercuryActivity(activity));
        }
      }
    });
  };
}

export function loadPreviousMessages(converstationId, lastActivity, spark) {
  return (dispatch) => {
    dispatch(updateConversationState({isLoadingHistoryUp: true}));
    spark.conversation.listActivities({
      conversationId: converstationId,
      limit: 20,
      maxDate: lastActivity.published
    })
    .then((activities) => {
      dispatch(addActivitiesToConversation(activities));
      dispatch(updateConversationState({isLoadingHistoryUp: false}));
    });
  };
}
