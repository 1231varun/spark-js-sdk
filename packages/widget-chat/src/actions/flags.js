export const ADD_FLAG_BEGIN = `ADD_FLAG_BEGIN`;
function addFlagBegin(activity) {
  return {
    type: ADD_FLAG_BEGIN,
    payload: {
      activity
    }
  };
}

export const ADD_FLAG = `ADD_FLAG`;
function addFlagSuccess(activity, flag) {
  return {
    type: ADD_FLAG,
    payload: {
      activity,
      flag
    }
  };
}

function addFlagError(activity, error) {
  return {
    type: ADD_FLAG,
    payload: {
      activity,
      error
    },
    error: true
  };
}

export const FLAGS_REQUEST_BEGIN = `FLAGS_REQUEST_BEGIN`;
function flagsRequestBegin() {
  return {
    type: FLAGS_REQUEST_BEGIN
  };
}

export const FLAGS_REQUEST = `FLAGS_REQUEST`;
function flagsRequestSuccess(flags) {
  return {
    type: FLAGS_REQUEST,
    payload: {
      flags
    }
  };
}

function flagsRequestError(error) {
  return {
    type: FLAGS_REQUEST,
    payload: error,
    error: true
  };
}

export const REMOVE_FLAG = `REMOVE_FLAG`;
function removeFlagBegin(flag) {
  return {
    type: REMOVE_FLAG,
    payload: {
      flag
    }
  };
}

function removeFlagError(error, flag) {
  return {
    type: REMOVE_FLAG,
    payload: {
      error,
      flag
    },
    error: true
  };
}

/**
 * Fetches all of the current user's flags
 *
 * @param {any} spark
 * @returns {function}
 */
export function fetchFlags(spark) {
  return (dispatch) => {
    dispatch(flagsRequestBegin());
    return spark.flag.list()
      .then((flags) =>
        dispatch(flagsRequestSuccess(flags))
      )
      .catch((error) =>
        dispatch(flagsRequestError(error))
      );
  };
}

/**
 * Flags a given activity. Updates state immediately then
 * adds flag details given from api
 *
 * @param {any} activity
 * @param {any} spark
 * @returns {function}
 */
export function flagActivity(activity, spark) {
  return (dispatch) => {
    dispatch(addFlagBegin(activity));
    return spark.flag.create(activity)
      .then((flag) =>
        dispatch(addFlagSuccess(activity, flag))
      )
      .catch((error) =>
        dispatch(addFlagError(activity, error))
      );
  };
}

/**
 * Removes a flag from the server. Updates the state immediately
 * but re-adds it if the delete fails
 *
 * @param {any} flag
 * @param {any} spark
 * @returns {function}
 */
export function removeFlagFromServer(flag, spark) {
  return (dispatch) => {
    dispatch(removeFlagBegin(flag));
    return spark.flag.delete(flag)
      .catch((error) => {
        dispatch(removeFlagError(error, flag));
      });
  };
}
