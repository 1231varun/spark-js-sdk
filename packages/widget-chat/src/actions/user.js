export const ADD_AVATAR_FOR_USER = `ADD_AVATAR_FOR_USER`;
function addAvatarForUserId(userId, avatar) {
  return {
    type: ADD_AVATAR_FOR_USER,
    userId,
    avatar
  };
}

export const BEGIN_FETCH_AVATAR_FOR_USER = `BEGIN_FETCH_AVATAR_FOR_USER`;
function beginFetchAvatarForUser(userId) {
  return {
    type: BEGIN_FETCH_AVATAR_FOR_USER,
    userId
  };
}

export const GET_CURRENT_USER = `GET_CURRENT_USER`;
export function getCurrentUser(user) {
  return {
    type: GET_CURRENT_USER,
    payload: {
      isFetching: false,
      user
    }
  };
}

export const GET_CURRENT_USER_BEGIN = `GET_CURRENT_USER_BEGIN`;
export function getCurrentUserBegin() {
  return {
    type: GET_CURRENT_USER_BEGIN,
    payload: {
      isFetching: true
    }
  };
}


export function fetchCurrentUser(spark) {
  return (dispatch) => {
    dispatch(getCurrentUserBegin());
    return spark.user.get()
      .then((user) => {
        dispatch(getCurrentUser(user));
      });
  };
}

export function fetchAvatarForUserId(userId, spark) {
  return (dispatch) => {
    dispatch(beginFetchAvatarForUser(userId));
    return spark.avatar.retrieveAvatarUrl(userId)
      .then((avatarUrl) =>
        dispatch(addAvatarForUserId(userId, avatarUrl))
      );
  };
}
