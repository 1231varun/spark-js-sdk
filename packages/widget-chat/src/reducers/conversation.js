/* eslint complexity: ["error", 10] */
// Refactoring and splitting up reducer in next feature
import _ from 'lodash';

import {
  ADD_ACTIVITIES_TO_CONVERSATION,
  CREATE_CONVERSATION,
  RECEIVE_CONVERSATION,
  RECEIVE_MERCURY_ACTIVITY,
  RECEIVE_MERCURY_COMMENT,
  UPDATE_CONVERSATION_STATE,
  UPDATE_MERCURY_STATE
} from '../actions/conversation';

const filteredActivities = [`delete`];

function filterActivity(activity) {
  return filteredActivities.indexOf(activity.verb) === -1;
}

function sortActivityByTime(activities) {
  activities = _.uniqBy(activities, `id`);
  return _.sortBy(activities, [`published`]);
}


export default function reduceConversation(state = {
  activities: [],
  id: null,
  participants: [],
  isFetching: false,
  isLoaded: false,
  isLoadingHistoryUp: false,
  mercuryState: {
    isListening: false
  }
}, action) {
  switch (action.type) {
  case ADD_ACTIVITIES_TO_CONVERSATION: {
    const activities = [...action.payload.activities, ...state.activities];
    return Object.assign({}, state, {
      activities: sortActivityByTime(activities)
    });
  }

  case CREATE_CONVERSATION: {
    return Object.assign({}, state, {
      isFetching: true
    });
  }

  case RECEIVE_CONVERSATION: {
    const activities = action.payload.conversation.activities.items.filter(filterActivity);

    return Object.assign({}, state, {
      activities,
      isFetching: false,
      isLoaded: true,
      id: action.payload.conversation.id,
      participants: action.payload.conversation.participants.items
    });
  }

  case RECEIVE_MERCURY_ACTIVITY: {
    let activities = state.activities;
    if (action.payload.activity.verb === `delete`) {
      // Find activity that is being deleted and change it to a tombstone
      const deletedId = action.payload.activity.object.id;
      activities = state.activities.map((activity) => {
        if (activity.id === deletedId) {
          return Object.assign({}, activity, {
            verb: `tombstone`
          });
        }
        return activity;
      });
    }
    return Object.assign({}, state, {
      activities: [...activities]
    });
  }

  case RECEIVE_MERCURY_COMMENT: {
    const activities = state.activities;
    const activity = action.payload.activity;
    return Object.assign({}, state, {
      activities: [...activities, activity]
    });
  }

  case UPDATE_CONVERSATION_STATE: {
    return Object.assign({}, state, action.payload.conversationState);
  }

  case UPDATE_MERCURY_STATE: {
    return Object.assign({}, state, {
      mercuryState: action.payload.mercuryState
    });
  }

  default: {
    return state;
  }
  }
}
