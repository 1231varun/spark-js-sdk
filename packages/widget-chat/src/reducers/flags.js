import {
  ADD_FLAG,
  BEGIN_RECEIVE_FLAGS,
  RECEIVE_FLAGS,
  REMOVE_FLAG
} from '../actions/flags';

function mapFlag(flag) {
  return {
    id: flag.id,
    url: flag.url,
    activityUrl: flag[`flag-item`]
  };
}

export default function conversation(state = {
  flags: [],
  hasFetched: false,
  isFetching: false
}, action) {
  switch (action.type) {

  case ADD_FLAG: {
    const flag = mapFlag(action.flag);
    return Object.assign({}, state, {
      flags: [...state.flags, flag]
    });
  }

  case BEGIN_RECEIVE_FLAGS: {
    return Object.assign({}, state, {
      isFetching: true
    });
  }

  case RECEIVE_FLAGS: {
    const flags = action.flags.map(mapFlag);
    return Object.assign({}, state, {
      hasFetched: true,
      isFetching: false,
      flags: [...flags]
    });
  }

  case REMOVE_FLAG: {
    const flags = state.flags.filter((flag) => flag.id !== action.flag.id);
    return Object.assign({}, state, {
      flags: [...flags]
    });
  }

  default: {
    return state;
  }
  }
}
