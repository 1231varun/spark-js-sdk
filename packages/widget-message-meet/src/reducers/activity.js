import {Map, OrderedMap} from 'immutable';

import {
  ADD_FILES_TO_ACTIVITY,
  RESET_ACTIVITY,
  REMOVE_FILE_FROM_ACTIVITY,
  SAVE_SHARE_ACTIVITY,
  SUBMIT_ACTIVITY_START,
  UPDATE_ACTIVITY_STATUS,
  UPDATE_ACTIVITY_TEXT
} from '../actions/activity';

const initialState = new Map({
  inFlightActivity: false,
  status: new Map({
    isSending: false,
    isTyping: false
  }),
  files: new OrderedMap(),
  text: ``
});

// eslint-disable-reason lots of actions for activities
// eslint-disable-next-line complexity
export default function reduceActivity(state = initialState, action) {
  switch (action.type) {
  case ADD_FILES_TO_ACTIVITY: {
    const files = action.payload.files.reduce((o, currentFile) => {
      o[currentFile.id] = currentFile;
      return o;
    }, {});

    return state.mergeIn([`files`], files);
  }

  case REMOVE_FILE_FROM_ACTIVITY: {
    return state.deleteIn([`files`, action.payload.id]);
  }

  case RESET_ACTIVITY: {
    return initialState;
  }

  case SAVE_SHARE_ACTIVITY: {
    return state.set(`shareActivity`, action.payload.shareActivity);
  }

  case SUBMIT_ACTIVITY_START: {
    return state
      // Store the activity as an in flight activity before sending
      .set(`inFlightActivity`, action.payload.activity)
      // Clear the text from the input
      .set(`text`, ``)
      .setIn([`status`, `isSending`], true);
  }

  case UPDATE_ACTIVITY_STATUS:
    return state.mergeDeepIn([`status`], action.payload.status);

  case UPDATE_ACTIVITY_TEXT:
    return state.set(`text`, action.payload.text);

  default:
    return state;
  }
}
