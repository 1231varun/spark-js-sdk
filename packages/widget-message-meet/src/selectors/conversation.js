import {createSelector} from 'reselect';
import _ from 'lodash';

const getActivities = (state) => state.conversation.activities.toArray(); // eslint-disable-line func-style
const getParticipants = (state) => state.conversation.participants; // eslint-disable-line func-style

export const getMostRecentActivity = createSelector(
  getActivities,
  (activities) => _.last(activities)
);

export const getMostRecentReadReceipts = createSelector(
  [getActivities, getParticipants],
  (activities, participants) => {
    const activity = _.last(activities);
    return participants.filter((participant) =>
      participant.roomProperties && participant.roomProperties.lastSeenActivityUUID === activity.id
    );
  }
);

/**
 * This creates an array to be used with the ActivityList component
 */
export const getActivityList = createSelector(
  [getActivities],
  (activities) => activities // Testing
);
