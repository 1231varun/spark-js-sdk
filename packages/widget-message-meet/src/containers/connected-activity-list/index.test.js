import React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import {initialState as conversation} from '../../reducers/conversation';
import {initialState as flags} from '../../reducers/flags';
import {initialState as user} from '../../reducers/user';

import ConnectedActivityList from '.';

const middlewares = [];
const mockStore = configureStore(middlewares);

const store = mockStore({
  conversation,
  flags,
  user
});

describe(`ConnectedActivityList`, () => {
  it(`renders properly`, () => {
    const component = renderer.create(
      <Provider store={store}>
        <ConnectedActivityList />
      </Provider>
    );

    const widget = component.toJSON();

    expect(widget).toMatchSnapshot();
  });
});
