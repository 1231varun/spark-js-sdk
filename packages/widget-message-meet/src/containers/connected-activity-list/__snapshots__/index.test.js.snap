exports[`ConnectedActivityList renders properly 1`] = `
<div
  className="activity-list" />
`;
